onnx/defs/shape_inference.h

```diff
@@ -613,6 +613,48 @@ inline void mergeInShapeInfo(
     mergeInShapeInfo(source.shape(), target);
 }
 
+// If two dimensions have different numerical values, return
+// false. Otherwise, true is returned.
+inline bool checkDimensionCompatibility(
+    const TensorShapeProto_Dimension& source_dim,
+    const TensorShapeProto_Dimension& target_dim) {
+  if (source_dim.has_dim_value() && target_dim.has_dim_value()) {
```

Comment:
Use switch statement instead of has_*() functions which are absent in proto3. You want this to run in ML.NET.

Reply:
I believe it is not a problem with "one of" fields … the has_* functions are generated even in proto3 for "one of", if I am not mistaken

Reply:
See https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#oneof-numeric-fields and https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#oneof-embedded-message-fields … both proto2 and proto3 support "has_*" for one of fields. It is only for primitive-type fields that are not "one of" we have the proto2/proto3 issue (IIUC).